Kafka Source/Sink added and Refactored Tasks

adamhass opened the pull request:
Two commits, don't squash them :) 

I've tried adding a ulimit increase to the ci yml, will see how that works with this PR. 

First Commit: Added Kafka integration. It's not at all optimized and it doesn't have well-defined behavior for committing reads/writes, but that will be added with State soon. The kafka integration requires the feature flag -kafka for it to be compiled/to run its tests. The tests are incomplete (no kafka mock to use) so no reason to add them to the CI pipeline yet, I'm thinking we add a kafka-integration test ci-pipeline when state is ready.

2nd Commit: Refactoring of our tasks so that we reduce code-duplication. I've called the new component "Node" but I'm not sure it's a good name? No need to rush to merging this PR. The point of this refactor is that the next steps of adding Epochs, separating input channels, buffers, etc. will be a lot easier and cleaner with this refactor since we decouple message handling from the tasks themselves.

Max-Meldrum commented:
Nice work! 😄 

Some cleanup would be good though, will comment on them.

Max-Meldrum commented on execution-plane/arcon/src/streaming/source/kafka.rs:
```diff
@@ -1,41 +1,281 @@
+    OUT: 'static + ArconType + DeserializeOwned,
+{
+    fn handle(&mut self, event: ControlEvent) -> () {
+        match event {
```
if let ControlEvent::Start = event {
   self.receive();
}

That should work. As the ControlEvent enum doesnt not have PartialEq, it doesn't work with a simple ==
or if

adamhass replied:
Is there any reason to not use the match? I think it's a better pattern for the ControlEvent because we may want to add features using other ControlEvents.

Max-Meldrum commented on execution-plane/arcon/src/streaming/source/kafka.rs:
```diff
@@ -1,41 +1,281 @@
+    }
+
+    #[test]
+    fn kafka_source() -> Result<()> { 
```
This test does not look like it is doing anything?
And it is sleeping 10 seconds

adamhass replied:
I was using the test to run "manual inspection tests" and figured it doesn't hurt to leave them as they were, useful for further development of them until we have a Kafka-mock or CI set-up with proper tests. It's not running when calling "cargo test" because it requires the feature flag to be added explicitly to the build/test command.

Max-Meldrum commented on execution-plane/arcon/src/streaming/source/kafka.rs:
```diff
@@ -1,41 +1,281 @@
+        y: f32,
+    }
+    // JSON Example: {"id":1, "attribute":-13,"location":{"x":0.14124,"y":5882.231}}
+    mod sink {
```
Is this needed? is it possible to use the Debug sink instead?

adamhass replied:
Yeah I noticed this as well while working on the next commit. Already working on changing all the tests to use the DebugSink 👍

Max-Meldrum commented on execution-plane/arcon/src/streaming/sink/kafka.rs:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    #[test]
+    fn kafka_sink() -> Result<()> { 
```
Same here. Is this test doing anything?

Max-Meldrum commented on execution-plane/arcon/src/streaming/sink/kafka.rs:
```diff
@@ -0,0 +1,169 @@
+    IN: 'static + ArconType,
+{
+    fn handle(&mut self, event: ControlEvent) -> () {
+        match event {
```
unnecessary match and error log.
if let ..  {
}

Max-Meldrum commented on execution-plane/arcon/src/streaming/task/filter.rs:
```diff
@@ -29,60 +19,89 @@ where
+        Ok(Vec::new())
+    }
+
+    fn handle_element(&mut self, element: ArconElement<IN>) -> ArconResult<Vec<ArconEvent<IN>>> {
```
utilising the ArconResult

`let result = self.run_udf(&(element.data))?;`
`if result == 1 {
   Ok(vec!(ArconEvent::Element(element))
} else {
   Ok(Vec::new())
}
`

Max-Meldrum commented on execution-plane/arcon/src/streaming/task/filter.rs:
```diff
@@ -29,60 +19,89 @@ where
+        let expected: Vec<i32> = vec![6];
+        let mut results = Vec::new();
+        for r in result_vec {
+            match r {
```
using if let here to avoid unnecessary  empty ` {} ` cases

Max-Meldrum commented on execution-plane/arcon/src/streaming/task/flatmap.rs:
```diff
@@ -35,66 +27,91 @@ where
+    }
+
+    fn handle_element(&mut self, element: ArconElement<IN>) -> ArconResult<Vec<ArconEvent<OUT>>> {
+        match self.run_udf(&(element.data)) {
```
same here with `let result = ....?;`

Max-Meldrum commented on execution-plane/arcon/src/streaming/task/flatmap.rs:
```diff
@@ -35,66 +27,91 @@ where
+        let expected: Vec<i32> = vec![6, 7, 8, 9, 10];
+        let mut result_vec = Vec::new();
+
+        match exec_result {
```
using if let here to avoid unnecessary  empty ` {} ` cases

Max-Meldrum commented on execution-plane/arcon/src/streaming/task/map.rs:
```diff
@@ -35,61 +26,89 @@ where
+    IN: 'static + ArconType,
+    OUT: 'static + ArconType,
+{
+    fn handle_element(&mut self, event: ArconElement<IN>) -> ArconResult<Vec<ArconEvent<OUT>>> {
```
remove commented debug statement and use the `let result = ...?;` approach

Max-Meldrum commented on execution-plane/arcon/src/streaming/task/map.rs:
```diff
@@ -35,61 +26,89 @@ where
+        let expected: Vec<i32> = vec![16,17];
+        let mut results = Vec::new();
+        for r in result_vec {
+            match r {
```
using if let here to avoid unnecessary empty {} cases

Max-Meldrum commented on execution-plane/arcon/src/streaming/task/node.rs:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+    fn output_event(&mut self, event: ArconEvent<OUT>) -> () {
+        match self.out_channels.output(event, &self.ctx.system()) {
```
unnecessary match statement.   doing a `if let Err(e) = .... {  error!() }` would be better

Max-Meldrum commented on execution-plane/arcon/src/streaming/task/node.rs:
```diff
@@ -0,0 +1,111 @@
+        OUT: 'static + ArconType,
+{
+    fn handle(&mut self, event: ControlEvent) -> () {
+        match event {
```
unnecessary match statement

adamhass commented:
Added a 3rd commit. Adds the ArconMessage class, an envelope for storing sender information.
Updated codegen to match the new structure.
Restructured the protobuff messages.
A few fix-ups of the changes requested.
